File: bibgraph/pipeline.py
"""End-to-end PDF -> structured JSON ingestion pipeline.

    pdf
     ├─ PyMuPDF        -> text-layer probe (OCR auto-detect) + link annotations
     ├─ MinerU (VLM)   -> content_list.json + middle.json + images/
     ├─ structure      -> section tree + floats
     ├─ references     -> structured bibliography
     ├─ annotate       -> inline [[cite:..]] / [[xref:..]] tokens (regex+links)
     └─ symbols        -> lightweight symbol inventory
    -> Document -> <out_dir>/<stem>.json
"""

from __future__ import annotations

import json
import logging
from pathlib import Path

from .config import PipelineConfig
from .mineru_client import MineruClient, MineruResult
from .pdf_links import (PdfLinks, bbox_1000_to_frac, extract_links,
                        has_text_layer, overlap_fraction)
from .schema import (Document, EquationBlock, FigureBlock, ListBlock,
                     Paragraph, RichText)
from .ingest.annotate import apply_matches
from .ingest.citations import (ReferenceResolver, detect_citations,
                               enrich_citations_with_links)
from .ingest.crossrefs import XrefIndex, detect_crossrefs, enrich_crossrefs_with_links
from .ingest.references import parse_references
from .ingest.structure import build_structure
from .ingest.symbols import extract_symbols, fill_equation_symbols

log = logging.getLogger("bibgraph.pipeline")


def ingest_pdf(pdf_path: str | Path, out_dir: str | Path | None = None,
               config: PipelineConfig | None = None,
               use_mineru_cache: bool = True,
               write_json: bool = True) -> Document:
    config = config or PipelineConfig()
    pdf_path = Path(pdf_path)
    if not pdf_path.exists():
        raise FileNotFoundError(pdf_path)
    out_dir = Path(out_dir) if out_dir else (pdf_path.parent / pdf_path.stem)
    out_dir.mkdir(parents=True, exist_ok=True)

    # 1) OCR auto-detect (only when caller didn't force it)
    if config.mineru.is_ocr is None:
        has_text = has_text_layer(pdf_path)
        config.mineru.is_ocr = not has_text
        log.info("OCR auto-detect: text_layer=%s -> is_ocr=%s",
                 has_text, config.mineru.is_ocr)

    # 2) MinerU extraction (cached if available)
    client = MineruClient(config.mineru)
    mineru = client.extract(pdf_path, out_dir,
                            poll_interval=config.poll_interval,
                            poll_timeout=config.poll_timeout,
                            use_cache=use_mineru_cache)

    # 3) PDF hyperlink annotations (hybrid resolution)
    pdf_links: PdfLinks | None = None
    if config.use_pdf_links:
        try:
            pdf_links = extract_links(pdf_path)
        except Exception as e:                       # never fail the whole run
            log.warning("Link extraction failed (%s); regex-only resolution", e)

    doc = build_document(mineru, pdf_path, pdf_links, config)

    if write_json:
        out_json = out_dir / f"{pdf_path.stem}.json"
        out_json.write_text(
            json.dumps(doc.to_dict(config.compact_json), ensure_ascii=False,
                       indent=2), "utf-8")
        log.info("Wrote %s", out_json)
        doc.meta["output_path"] = str(out_json)
    return doc


def build_document(mineru: MineruResult, pdf_path: Path,
                   pdf_links: PdfLinks | None,
                   config: PipelineConfig) -> Document:
    """Assemble a :class:`Document` from already-extracted artifacts.

    Separated from :func:`ingest_pdf` so the offline parsing stages can be
    exercised without calling the MinerU API.
    """
    structure = build_structure(mineru.content_list)
    references = parse_references(structure.ref_text_items)

    n_pages = pdf_links.n_pages if pdf_links else _max_page(mineru.content_list)
    doc = Document(
        doc_id=pdf_path.stem,
        source={"type": "pdf", "path": str(pdf_path), "filename": pdf_path.name,
                "n_pages": n_pages},
        meta={"title": structure.title_guess,
              "mineru": {"model_version": config.mineru.model_version,
                         "language": config.mineru.language,
                         "is_ocr": config.mineru.is_ocr,
                         "batch_id": mineru.batch_id}},
        structure=structure.sections,
        references=references,
    )

    resolver = ReferenceResolver(references)
    xindex = XrefIndex(doc)
    _annotate_document(doc, resolver, xindex, pdf_links,
                       use_links=config.use_pdf_links)

    fill_equation_symbols(doc)
    doc.symbols = extract_symbols(doc)
    return doc


def _annotate_document(doc: Document, resolver: ReferenceResolver,
                       xindex: XrefIndex, pdf_links: PdfLinks | None,
                       use_links: bool) -> None:
    for block in doc.iter_blocks():
        holders = _annotatable(block)
        if not holders:
            continue
        block_links = _links_for_block(block, pdf_links) if use_links else []
        for holder in holders:
            if not holder.text:
                continue
            cmatches = detect_citations(holder.text, resolver)
            xmatches = detect_crossrefs(holder.text, xindex)
            if block_links:
                enrich_citations_with_links(cmatches, block_links, resolver)
                enrich_crossrefs_with_links(xmatches, block_links, xindex)
            new_text, cites, xrefs = apply_matches(holder.text, cmatches + xmatches)
            holder.text = new_text
            holder.citations = cites
            holder.crossrefs = xrefs


def _annotatable(block) -> list:
    """Return the rich-text holders of a block (each has .text/.citations/.crossrefs)."""
    if isinstance(block, Paragraph):
        return [block]
    if isinstance(block, ListBlock):
        return list(block.items)
    if isinstance(block, EquationBlock):
        return []
    cap = getattr(block, "caption", None)
    return [cap] if isinstance(cap, RichText) else []


def _links_for_block(block, pdf_links: PdfLinks | None) -> list:
    if pdf_links is None or block.page_idx is None:
        return []
    bbox = bbox_1000_to_frac(getattr(block, "bbox", None))
    if bbox is None:
        return []
    return [ln for ln in pdf_links.links_on_page(block.page_idx)
            if overlap_fraction(ln.rect, bbox) > 0.5]


def _max_page(content_list: list[dict]) -> int:
    pages = [it.get("page_idx") for it in content_list
             if isinstance(it.get("page_idx"), int)]
    return (max(pages) + 1) if pages else 0

File: packages/core/src/pipelines/pdf/pipeline.ts
/**
 * End-to-end PDF -> structured JSON ingestion pipeline (`bibgraph/pipeline.py`).
 *
 *     pdf
 *      ├─ mupdf          -> text-layer probe (OCR auto-detect) + link annotations [port]
 *      ├─ MinerU (VLM)   -> content_list.json + middle.json + images/             [port]
 *      ├─ structure      -> section tree + floats
 *      ├─ references     -> structured bibliography
 *      ├─ textfix        -> repair VLM '?'-gaps from the PDF text layer           [port]
 *      └─ annotate       -> inline [[cite:..]] / [[xref:..]] tokens (regex+links)
 *     -> Document -> <out_dir>/<stem>.json
 *
 * The offline stages (structure / references / textfix / annotate / document
 * assembly) are the shared documents-domain pieces composed by
 * {@link buildDocument}; what lives here is the orchestration `ingest_pdf` adds
 * around them: OCR auto-detect, the MinerU call (cache-aware), the never-fail
 * link-extraction guard, and JSON emission.
 */

import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { basename, dirname, join, parse } from "node:path";
import type { Document } from "@argelanderspace/contracts";
import { buildDocument, documentToJson } from "../../documents/document.js";
import type { PdfLinks } from "../../documents/pdf-links.js";
import type { PdfTextProvider } from "../../documents/textfix.js";
import {
  DEFAULT_PDF_CONFIG,
  type PdfMineruConfig,
  type PdfPipelineConfig,
  type PdfPipelineConfigOverrides,
  type PdfPipelinePorts,
} from "./ports.js";

export interface IngestPdfOptions {
  /** Output directory (Python `out_dir`, default `<pdf dir>/<pdf stem>`). */
  outDir?: string;
  /** `PipelineConfig` overrides; defaults come from `bibgraph/config.py`. */
  config?: PdfPipelineConfigOverrides;
  /** Reuse a previous MinerU unzip under `<out_dir>/mineru` (Python `use_mineru_cache`). */
  useMineruCache?: boolean;
  /** Write `<out_dir>/<stem>.json` (Python `write_json`). */
  writeJson?: boolean;
  /**
   * Coarse progress sink (Stage 3 / MS3): one line per stage transition
   * (MinerU extraction / link harvest / document build / JSON write), plus the
   * MinerU port's own extraction-state transitions. The upload job wires this
   * to the job runner's `report`.
   */
  onProgress?: (message: string) => void;
}

/** `ingest_pdf(pdf_path, out_dir, config, use_mineru_cache, write_json)`. */
export async function ingestPdf(
  pdfPath: string,
  ports: PdfPipelinePorts,
  opts: IngestPdfOptions = {}
): Promise<Document> {
  // The Python pipeline mutates config.mineru.is_ocr during auto-detect; resolve
  // into a local copy instead (caller-visible side effect dropped, outputs equal).
  const mineruConfig: PdfMineruConfig = { ...DEFAULT_PDF_CONFIG.mineru, ...opts.config?.mineru };
  const config: PdfPipelineConfig = {
    ...DEFAULT_PDF_CONFIG,
    ...opts.config,
    mineru: mineruConfig,
  };
  if (!existsSync(pdfPath)) {
    throw new Error(`PDF not found: ${pdfPath}`);
  }
  const stem = parse(pdfPath).name;
  const outDir = opts.outDir ?? join(dirname(pdfPath), stem);
  mkdirSync(outDir, { recursive: true });

  // 1) OCR auto-detect (only when the caller didn't force it)
  if (mineruConfig.isOcr === null || mineruConfig.isOcr === undefined) {
    mineruConfig.isOcr = !ports.links.hasTextLayer(pdfPath);
  }

  // 2) MinerU extraction (cached if available)
  opts.onProgress?.("MinerU extraction");
  const mineru = await ports.mineru(pdfPath, outDir, {
    config: mineruConfig,
    pollInterval: config.pollInterval,
    pollTimeout: config.pollTimeout,
    useCache: opts.useMineruCache ?? true,
    onProgress: opts.onProgress,
  });

  // 3) PDF hyperlink annotations (hybrid resolution) — a failed extraction never
  // fails the whole run; resolution degrades to regex-only.
  let pdfLinks: PdfLinks | null = null;
  if (config.usePdfLinks) {
    opts.onProgress?.("Extracting PDF links");
    try {
      pdfLinks = ports.links.extract(pdfPath);
    } catch {
      pdfLinks = null;
    }
  }

  // textfix opens the PDF itself in Python and records zeroed stats when that
  // open fails; buildDocument records the same zeroed stats without a provider.
  let pdfText: PdfTextProvider | undefined;
  if (config.useTextfix) {
    try {
      pdfText = ports.openText(pdfPath);
    } catch {
      pdfText = undefined;
    }
  }

  opts.onProgress?.("Building document");
  const doc = buildDocument(
    mineru,
    { stem, path: pdfPath, filename: basename(pdfPath) },
    pdfLinks,
    {
      usePdfLinks: config.usePdfLinks,
      useTextfix: config.useTextfix,
      mineru: {
        modelVersion: mineruConfig.modelVersion,
        language: mineruConfig.language,
        isOcr: mineruConfig.isOcr,
      },
      pdfText,
    }
  );

  if (opts.writeJson ?? true) {
    opts.onProgress?.("Writing document JSON");
    const outJson = join(outDir, `${stem}.json`);
    writeFileSync(
      outJson,
      JSON.stringify(documentToJson(doc, config.compactJson), null, 2),
      "utf-8"
    );
    // Python sets meta.output_path on the returned doc *after* serialization.
    if (doc.meta) doc.meta.output_path = outJson;
  }
  return doc;
}
